packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3256 @@
+///
+/// This is a Material Design menu bar that resides above the main body of an
+/// application that defines a menu system for invoking callbacks or firing
+/// [Intent]s in response to user selection of the menu item.
+///
+/// The menu can be navigated by the user using the arrow keys, and can be
+/// dismissed using the escape key, or by clicking away from the menu item
+/// (anywhere that is not a part of the menu bar or cascading menus). Once a
+/// menu is open, the menu hierarchy can be navigated by hovering over the menu
+/// with the mouse.
+///
+/// Menu items can have a [SingleActivator] or [CharacterActivator]  assigned to
+/// them as their [MenuBarItem.shortcut], so that if the shortcut key sequence
+/// is pressed, the menu item corresponding to that shortcut will be selected.
+/// Shortcuts must be unique in the ambient [ShortcutRegistry].
+///
+/// Selecting a menu item causes the [MenuBarItem.onSelected] callback to be
+/// called.
+///
+/// When a menu item with a submenu is clicked on, it toggles the visibility of
+/// the submenu. When the menu item is hovered over, the submenu will open after
+/// a slight delay, and hovering over other items will close that menu and open
+/// the newly hovered one. When those occur, [MenuBarMenu.onOpen], and
+/// [MenuBarMenu.onClose] are called, respectively.
+///
+/// {@tool dartpad}
+/// This example shows a [MenuBar] that contains a single top level menu,
+/// containing three items for "About", a checkbox menu item for showing a
+/// message, and "Quit". The items are identified with an enum value.
+///
+/// ** See code in examples/api/lib/material/menu_bar/menu_bar.0.dart **
+/// {@end-tool}
+///
+/// {@tool sample}
+/// This example shows a [MenuBar] that contains a simple menu for a desktop
+/// application. It is set up to be adaptive, so that on macOS it will use the
+/// system menu bar, and on other systems will use a Material [MenuBar].
+///
+/// On macOS, it will add the "About" and "Quit" system-provided menu items.
+///
+/// The menu items are all identified by an enum value (`MenuSelection`).
+///
+/// ** See code in examples/api/lib/material/menu_bar/menu_bar.1.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [MenuBarMenu], a menu item which manages a submenu.
+///  * [MenuItemGroup], a menu item which collects its members into a group
+///    separated from other menu items by a divider.
+///  * [MenuBarItem], a leaf menu item which displays the label, an optional
+///    shortcut label, and optional leading and trailing icons.
+///  * [MenuBarController], a class that allows closing of menus from outside of
+///    the menu bar.
+///  * [PlatformMenuBar], which creates a menu bar that is rendered by the host
+///    platform instead of by Flutter (on macOS, for example).
+///  * [ShortcutRegistry], a registry of shortcuts that apply for the entire
+///    application, used by the `MenuBar` to register its shortcuts.
+class MenuBar extends PlatformMenuBar {
```

Comment:
Wondering if this file should be re-organized a little bit. The thing that most people coming to this file will likely care about, doesn't appear until 1157 lines into the file...

Reply:
Once I move the _MenuController code into the state, it'll be close to the top.